Evan Forletta

Fall, 2018





# Part 2:  Functions and Control Statements

A distinguishing property of *programming* languages is that the programmer can create their own *functions*.  Creating a *function* is like teaching the computer a new trick.  Typically a function will receive some data as *input*, will perform an *algorithm* involving the input data, and will *output* data when the algorithm terminates.  

In this part, we explore Python functions.  We also explore control statements, which allow a program to behave in different ways for different inputs.  We also introduce the *while loop*, a loop whose repetition can be more carefully controlled than a for loop.  As an application of these techniques, we implement the Euclidean algorithm as a Python function in a few ways, to effectively find the GCD of integers.

At the end, you will be prepared to explore the Collatz conjecture.

## Table of Contents

- [Getting started with Python functions](#functions)
- [Control statements](#controls)
- [While loops and implementation of the Eucidean algorithm](#while)


<a id='functions'></a>

## Getting started with Python functions

A *function* in Python is a construction which takes input data, performs some actions, and outputs data.  It is best to start with a few examples and break down the code.  Here is a function `square`.  Run the code as usual by pressing *shift-Enter* when the code block is selected.

In [1]:
def square(x):
    answer = x * x
    return answer

When you run the code block, you probably didn't see anything happen.  But you have effectively taught your computer a new trick, increasing the vocabulary of commands it understands through the Python interpreter.  Now you can use the `square` command as you wish.

In [2]:
square(12)

144

In [3]:
square(1.5)

2.25

Let's break down the syntax of the *function declaration*, line by line.

```python
def square(x):
    answer = x * x
    return answer
```

The first line begins with the Python reserved word `def`.  (So don't use `def` as a variable name!).  The word `def` stands for "define" and it defines a function called `square`.  After the function name `square` comes parentheses `(x)` containing the **argument** `x`.  The *arguments* or *parameters* of a function refer to the input data.  Even if your function has no arguments, you need parentheses.  The argument `x` is used to name whatever number is input into the `square` function.  

At the end of the function declaration line is a colon `:` and the following two lines are indented.  As in the case of for loops, the colon and indentation are signals of *scope*.  Everything on the indented lines is considered within the *scope of the function* and is carried out when the function is used later.

The second line `answer = x * x` is the beginning of the scope of the function.  It declares a variable `answer` and sets the value to be `x * x`.  So if the argument `x` is 12, then `answer` will be set to 144.  The variable `answer`, being declared within the scope of the function, will not be accessible outside the scope of the function.  It is called a **local variable**.

The last line `return answer` contains the Python reserved word `return`, which terminates the function and outputs the value of the variable `answer`.  So when you apply the function with the command `square(1.5)`, the number `1.5` is `passed` as the argument `x`, and `answer` is `2.25`, and that number `2.25` becomes the output.

A function does not have to return a value.  Some functions might just provide some information.  Here is a function which displays the result of division with remainder as a sentence with addition and multiplication.

In [4]:
def display_divmod(a,b):
    quotient = a // b # Integer division
    remainder = a % b #
    print("{} = {} ({}) + {}".format(a,quotient,b,remainder))

In [5]:
display_divmod(23,5)

23 = 4 (5) + 3


Notice that this function has no `return` line.  The function terminates automatically at the end of its scope.

The function also uses Python's **string formatting**.  This has changed between Python 2.x and 3.x, and this notebook uses Python 3.x syntax.

String formatting allows you to insert placeholders like `{}` within a string, and later fill those places with a list of things.  

In [6]:
print("%d : %d" % (1,2))

1 : 2


In [7]:
print("My favorite number is {}".format(17))  # The .format "method" substitutes 17 for {}

My favorite number is 17


In [8]:
print("{} + {} = {}".format(13,12,13+12))

13 + 12 = 25


The `format` command is an example of a **string method**.  It has the effect of replacing all placeholders `{}` by the its inputs, in sequence.  There is an intricate syntax for these placeholders, to allow one to match placeholders with values in different orders, and to format different kinds of values.  Here is the [official reference for string formatting in Python 3.x](https://docs.python.org/3/library/string.html#formatstrings).  We will only use the most basic features, exhibited below.

In [9]:
print ("The number {} comes before {}.".format(1,2)) # This should be familiar.
print ("The number {1} comes before {0}.".format(1,2)) # What happens?
print ("The number {1} comes before {1}.".format(1,2)) # Got it now?

The number 1 comes before 2.
The number 2 comes before 1.
The number 2 comes before 2.


By placing a number in the placeholder, like `{1}`, one can fill in the placeholders with the values in a different order, or repeat the same value.  The format method takes multiple parameters, and they are numbered:  parameter 0, parameter 1, parameter 2, etc..  So the placeholder `{1}` will be replaced by the second parameter (parameter 1).  It's confusing at first, but Python almost always starts counting at zero.

In [11]:
print("pi is approximately {0}".format(3.14159265))
print("pi is approximately {0:f}".format(3.14159265)) # The "f" in "0:f" formats the float.
print("pi is approximately {0:0.3f}".format(3.14159265)) # Choose 3 digits of precision.

pi is approximately 3.14159265
pi is approximately 3.141593
pi is approximately 3.142


If you give some information about how the placeholder is being used, the format method will format things more nicely for printing.  The placeholder `{0:f}` will be replaced by parameter 0, and it will be formatted in a way that is nice for floats (hence the `f`).  Don't try formatting things outside of their type!

In [12]:
print("{:d} is a pretty big integer.".format(2**100)) # d is the formatting code for integers.
print("{:f} is an integer, formatted like a float.".format(2**100))
print("{:f} is a float, of course.".format(1/7))
print("{:s} is a string.".format('Hi there!')) # s is the formatting code for strings.
print("{:d} will give us an error message.".format(1/7))


1267650600228229401496703205376 is a pretty big integer.
1267650600228229401496703205376.000000 is an integer, formatted like a float.
0.142857 is a float, of course.
Hi there! is a string.


ValueError: Unknown format code 'd' for object of type 'float'

In [110]:
from math import sqrt  # Make sure the square root function is loaded.
print("The square root of {0:d} is about {1:f}.".format(1000, sqrt(1000)))

The square root of 1000 is about 31.622777.


### Exercises

1.  What are the signals of scope in Python?

2.  Write a function called area_circle, which takes one argument radius. The function should return the area of the circle, as a floating point number. Then add one line to the function, using string formatting, so that it additionally prints a helpful sentence of the form "The area of a circle of radius 1.0 is 3.14159." (depending on the radius and the area it computes).

3. `format` is an example of a "string method".  Another neat one is `replace`.  Try `"Python".replace("yth","arag")` to see what it does.  

4.  Try the formatting codes `%` and `E` (instead of `f`) for a floating point number.  What do they do?

5. Can you think of a reason you might want to have a function with *no* arguments?

1. There are two 'scopes' in python: global and local variables. Variables defined outside of functions are global, and can be ascessed 'as is' inside function definitions. Variables defined inside function definitions are local to that function however. A variable declaration `x=1` inside a function cannot be accesed in another function (unless x is already global). There's a case where a variable name can be both global *and* local. Specifically, if the variable name is also the name of a function's argument. For example
```python
x = 3
def square(x):
     return x**2
```
The code `square(7)` will return `49` not `9`.


2. Two function definitions:
```python
def area_circle(r):
     return 3.1415926536*(r**2)
```
```python
# Or also:
def area_circle(r):
     area = 3.1415926536*(r**2)
     print("The area of a circle of radius {} is {}".format(r,area))
```

3. "Paragon," that's great!


4. For a float x the code `%` prints the number x as... well, a percent. Specifically it multiplies x by 100 and then appends the string "%". Neat stuff. The code `E` prints the number in scientific notation, and computes exponents and inserts the string "E+" or "E-" to do so. Also cool!


5. Yeah, if you just want to do some task that doesn't depend on any specific input. Perhaps a function `clean_up()`, or `print_time()`, or something like that.

#### Scratch

In [16]:
a = 1
def p():
    print(a)
    
p()

1


In [17]:
def area_circle(r):
    return 3.1415926536*(r**2)

In [18]:
from dis import dis

In [19]:
dis(area_circle)

  2           0 LOAD_CONST               1 (3.1415926536)
              2 LOAD_FAST                0 (r)
              4 LOAD_CONST               2 (2)
              6 BINARY_POWER
              8 BINARY_MULTIPLY
             10 RETURN_VALUE


In [21]:
# dis doesn't work with built in functions
dis(print)

TypeError: don't know how to disassemble builtin_function_or_method objects

In [23]:
# Amazing
"Python".replace("yth","arag")

'Paragon'

In [26]:
number = 314159.26535897932
print("Check it: {:f}".format(number))
print("Check it: {:E}".format(number))
print("Check it: {:%}".format(number))

Check it: 314159.265359
Check it: 3.141593E+05
Check it: 31415926.535898%


<a id='controls'></a>

## Control statements

It is important for a computer program to behave differently under different circumstances.  The simplest control statements, `if` and its relative `else`, can be used to tell Python to carry out different actions depending on the value of a boolean variable.  The following function exhibits the syntax.

In [27]:
def is_even(n):
    if n%2 == 0:
        print("{} is even.".format(n))
        return True
    else:
        print("{} is odd.".format(n))
        return False

In [28]:
is_even(17)

17 is odd.


False

In [29]:
is_even(1000)

1000 is even.


True

The broad syntax of the function should be familiar.  We have created a function called `is_even` with one argument called `n`.  The body of the function uses the **control statement** `if n%2 == 0:`.  Recall that `n%2` gives the remainder after dividing `n` by `2`.  Thus `n%2` is 0 or 1, depending on whether `n` is even or odd.  Therefore the **boolean** `n%2 == 0` is `True` if `n` is even, and `False` if `n` is odd.

The next two lines (the first `print` and `return` statements) are within the **scope** of the `if <boolean>:` statement, as indicated by the colon and the indentation.  The `if <boolean>:` statement tells the Python interpreter to perform the statements within the scope if the boolean is `True`, and to ignore the statements within the scope if the boolean is `False`.

Putting it together, we can analyze the code.
```python
    if n%2 == 0:
        print("{} is even.".format(n))
        return True
```
If `n` is even, then the Python interpreter will print a sentence of the form `n is even`.  Then the interpreter will return (output) the value `True` and the function will terminate.  If `n` is odd, the Python interpreter will ignore the two lines of scope.

Often we don't just want Python to *do nothing* when a condition is not satisfied.  In the case above, we would rather Python tell us that the number is odd.  The `else:` control statement tells Python what to do in case the `if <boolean>:` control statement receives a `False` boolean.  We analyze the code
```python
    else:
        print("{} is odd.".format(n))
        return False
```
The `print` and `return` commands are within the scope of the `else:` control statement.  So when the `if` statement receives a false signal (the number `n` is odd), the program prints a sentence of the form `n is odd.` and then returns the value `False` and terminates the function.

The function `is_even` is a verbose, or "talkative" sort of function.  Such a function is sometimes useful in an interactive setting, where the programmer wants to understand everything that's going on.  But if the function had to be called a million times, the screen would fill with printed sentences!  In practice, an efficient and silent function `is_even` might look like the following.

In [30]:
def is_even(n):
    return (n%2 == 0)

In [31]:
is_even(17)

False

A `for` loop and an `if` control statement, used together, allow us to carry out a **brute force** search.  We can search for factors in order to check whether a number is prime.  Or we can look for solutions to an equation until we find one.

One thing to note:  the function below begins with a block of text between a triple-quote (three single-quotes when typing).  That text is called a **docstring** and it is meant to document what the function does.  Writing clear docstrings becomes more important as you write longer programs, collaborate with other programmers, and when you want to return months or years later to use a program again.  There are different style conventions for docstrings; for example, here are [Google's docstring conventions](https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments).  We take a less formal approach.

In [32]:
def is_prime(n):
    '''
    Checks whether the argument n is a prime number.
    Uses a brute force search for factors between 1 and n.
    '''
    for j in range(2,n):  # the list of numbers 2,3,...,n-1.
        if n%j == 0:  # is n divisible by j?
            print("{} is a factor of {}.".format(j,n))
            return False
    return True

An important note:  the `return` keyword **terminates** the function.  So as soon as a factor is found, the function terminates and outputs `False`.  If no factor is found, then the function execution survives past the loop, and the line `return True` is executed to terminate the function.

In [33]:
is_prime(91)

7 is a factor of 91.


False

In [34]:
is_prime(101)

True

Try the `is_prime` function on bigger numbers -- try numbers with 4 digits, 5 digits, 6 digits.  Where does it start to slow down?  Do you get any errors when the numbers are large?  Make sure to save your work first, just in case this crashes your computer!  



In [ ]:
# This dterminates right away because it's even
is_prime(100000000000000000000000000000000)

In [57]:
is_prime(100000000000000000000000000000001)

19841 is a factor of 100000000000000000000000000000001.


False

In [50]:
# Takes forever, KeyboardInterrupt
is_prime(18326287978278373842783872782872637388828872672132375627873)

KeyboardInterrupt: 

There are two limiting factors, which we study in more detail later.  These are **time** and **space** (your computer's memory space).  As the loop of `is_prime` goes on and on, it might take your computer a long time!  If each step of the loop takes only a nanosecond (1 billionth of a second), the loop would take about a second when executing `is_prime(1000000001)`.  If you tried `is_prime` on a much larger number, like `is_prime(2**101 - 1)`, the loop would take longer than the lifetime of the Earth.

The other issue that can arise is a problem with *space*.  In Python 3.x, the `range(2,n)` cleverly *avoids* storing all the numbers between `2` and `n-1` in memory.  It just remembers the endpoints, and how to proceed from one number to the next.  In the older version, Python 2.x, the range command `range(2,n)` would have tried to store the entire list of numbers `[2,3,4,...,n-1]` in the memory of your computer.  Your computer has some (4 or 8 or 16, perhaps) gigabytes of memory (RAM).  A gigabyte is a billion bytes, and a byte is enough memory to store a number between 0 and 255.  (More detail about this later!).  So a gigabyte will not even hold a billion numbers.  So our `is_prime` function would have led to memory problems in Python 2.x, but in Python 3.x we don't have to worry (for now) about space.

### Exercises

1.  Create a function `my_abs(x)` which outputs the absolute value of the argument `x`.  (Note that Python already has a built-in `abs(x)` function).  

2.  Modify the `is_prime` function so that it prints a message `Number too big` and returns `None` if the input argument is bigger than one million.  (Note that `None` is a Python reserved word.  You can use the one-line statement `return None`.)  

3.  Write a Python function `thrarity` which takes an argument `n`, and outputs the string `threeven` if `n` is a multiple of three, or `throdd` is `n` is one more than a multiple of three, or `thrugly` if `n` is one less than a multiple of three.  Example:  `thrarity(31)` should output `throdd` and `thrarity(44)` should output `thrugly`.  Hint:  study the `if`/`elif` syntax at [the official Python tutorial](https://docs.python.org/3/tutorial/controlflow.html#if-statements)

4.  Write a Python function `sum_of_squares(n)` which finds and prints a pair of natural numbers $x$, $y$, such that $x^2 + y^2 = n$.  The function should use a brute force search and return `None` if no such pair of numbers $x,y$ exists.  Explore which natural numbers can be expressed as sums of two squares... hint:  look at prime numbers first!

5.  Write a function `gamma(n)` which takes a positive integer n as input, and outputs the difference between the harmonic sum $\frac{1}{1} + \frac{1}{2} + \frac{1}{3} + \cdots + \frac{1}{n}$ and the natural logarithm $\log(n)$.  Use numpy to compute the logarithm, by using the command `from numpy import log` in a separate cell.  Approximate $\gamma(n)$ as $n \rightarrow \infty$.  How large does $n$ need to be to get five digits of precision on this limit?  Can you prove that the limit $\lim_{n \rightarrow \infty} \gamma(n)$ exists?

### Answers:

1. `my_abs(x)`:

```python
def my_abs(x):
    if x >= 0:
        return x
    else:
        return -x
```

2. `better_is_prime(n)`:
```python
def better_is_prime(n):

    # Protec against large numbers
    if x > 10**6:
        print("Number is too big!")
        return None

    # Loop through potential divisiors up to root n
    steps = int(sqrt(x))

    for d in range(2, steps+1):
        if x % d == 0:
            print("{} is a factor of {}.".format(d,n))
            return False

    return True
```


3. `thrarity(n)`:
```python
def thrarity(n):
    
    r = n%3 # This remainder will work even for negative numbers too!
    
    if r == 0:
        print("Threeven")
    elif r == 1:
        print("Throdd")
    else:
        print("THRUGLY")
```

4. `sum_of_squares(n)`: It looks like primes congruent to 1 (mod 4) are expressable as sums of squares!

```python
# The square testing function sucks for large numbers, so this will only be semi-accurate. 
# Works well for small numbers though!
def sum_of_squares(n):
    bound = int(sqrt(n))
    
    # Note we allow that 0 and 1 are valid squares
    for x in range(0, bound+1):
        # Solve for the would be y^2 term
        y_squared = n - x**2
        
        # Now test if y_squared anctually is a square
        root = sqrt(y_squared)
        root_int = int(root) 
        
        if root_int**2 == y_squared:
            # y is probably a square
            print("{} = {}^2 + {}^2".format(n, x, root_int))
    
    # We reached here and no solution found. Return None
    return None
```


5. `gamma(n)`:
```python
def gamma(n):
    H = 0
    for k in range(1,n+1):
        H += 1/k
        
    gamma = H - log(n)
    
    return gamma
```
This function gets gamma accurate to 5 decimal places arround 116000 itterations. Is that way to huge, or does it just go that slowly? The proof that $\lim_{n \to \infty} \gamma (n)$ converges is that it is monotonically increasing and bounded above. These come from that cool picture of $\int_1^\infty \frac{1}{x} dx$ and it's approximation by the harmonic sum.


### A bunch of scratch work

In [63]:
#  Use this space for your solutions to the questions.
def my_abs(x):
    if x >= 0:
        return x
    else:
        return -x

In [65]:
def better_is_prime(n):
    
    # Protec against large numbers
    if x > 10**6:
        print("Number is too big!")
        return None
    
    # Loop through potential divisiors up to root n
    steps = int(sqrt(x))
    
    for d in range(2, steps+1):
        if x % d == 0:
            print("{} is a factor of {}.".format(d,n))
            return False
    
    return True

In [73]:
def thrarity(n):
    
    r = n%3 # This remainder will work even for negative numbers too!
    
    if r == 0:
        print("Threeven")
    elif r == 1:
        print("Throdd")
    else:
        print("THRUGLY")

In [197]:
# The square testing function sucks for large numbers, so this will only be semi accurate. 
# Works well for small numbers though!
def sum_of_squares(n):
    bound = int(sqrt(n))
    
    # Note we allow that 0 and 1 are valid squares
    for x in range(0, bound+1):
        # Solve for the would be y^2 term
        y_squared = n - x**2
        
        # Now test if y_squared anctually is a square
        root = sqrt(y_squared)
        root_int = int(root) 
        
        if root_int**2 == y_squared:
            # y is probably a square
            print("{} = {}^2 + {}^2".format(n, x, root_int))
    
    # We reached here and no solution found. Return None
    return None

In [198]:
sum_of_squares(10)

10 = 1^2 + 3^2
10 = 3^2 + 1^2


In [130]:
big = 345334533248543767**2
root = sqrt(big)

root_sq = root**2

big - root_sq

1.8446744073709552e+19

In [141]:
sqrt(34453764567**2)**2

1.1870618928382647e+21

In [131]:
# This one fails a lot
def is_square(x):
    
    root = sqrt(x)
    
    if root**2 == x:
        return True
    else:
        return False

In [142]:
def is_square_2(x):
    
    root = sqrt(x)
    
    if x - root**2 == 0:
        return True
    else:
        return False

In [166]:
def is_square_3(x):
    root = sqrt(x)
    return int(root)**2 == x

In [184]:
is_square(25*25**8)

True

In [140]:
is_square(34453764567**2)

False

In [167]:
is_square_2(34453764567**2)

True

In [168]:
is_square_3(34453764567**2)

True

In [169]:
is_square(345334533248543767**2)

False

In [170]:
is_square_2(345334533248543767**2)

False

In [171]:
is_square_3(345334533248543767**2)

False

In [173]:
is_square(3445398765398765423456784567874567**2)

False

In [150]:
is_square_2(3445398765398765423456784567874567**2)

True

In [172]:
is_square_3(3445398765398765423456784567874567**2)

False

In [ ]:
one_hundred_and_ninety_primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151]
for p in one_hundred_and_ninety_primes:
    sum_of_squares(p)

In [203]:
from numpy import log

In [202]:
def gamma(n):
    H = 0
    for k in range(1,n+1):
        H += 1/k
        
    gamma = H - log(n)
    
    return gamma

In [220]:
gamma(116000)

0.5772199752400606

## Handling errors by raising exceptions.

In the previous batch of exercises, we tried to modify functions to be a bit more intelligent -- identifying when numbers were "too big" for example.  There's a professional way to handle these situations

<a id='while'></a>

## While loops and implementation of the Eucidean algorithm

We *almost* have all the tools we need to implement the Euclidean algorithm.  The last tool we will need is the **while loop**.  We have seen the *for loop* already, which is very useful for iterating over a range of numbers.  The Euclidean algorithm involves repetition, but there is no way to know in advance how many steps it will take.  The while loop allows us to repeat a process as long as a boolean value (sometimes called a **flag**) is True.  The following countdown example illustrates the structure of a while loop.

In [221]:
def countdown(n):
    current_value = n
    while current_value > 0:  # The condition (current_value > 0) is checked before every instance of the scope!
        print(current_value)
        current_value = current_value - 1
    print("Blastoff!")

In [222]:
countdown(10)

10
9
8
7
6
5
4
3
2
1
Blastoff!


The while loop syntax begins with `while <boolean>:` and the following indented lines comprise the scope of the loop.  If the boolean is `True`, then the scope of the loop is executed.  If the boolean is `True` again afterwards, then the scope of the loop is executed again.  And again and again and so on.

This can be a **dangerous process**!  For example, what would happen if you made a little typo and the last line of the while loop read `current_value = current_value + 1`?  The numbers would increase and increase... and the boolean `current_value > 0` would **always** be `True`.  Therefore the loop would never end.  Bigger and bigger numbers would scroll down your computer screen.  

You might panic under such a circumstance, and maybe turn your computer off to stop the loop.  Here is some advice for when your computer gets stuck in such a neverending loop:

1.  Back up your work often.  When you're programming, make sure everything else is saved just in case.
2.  Save your programming work (use "Save and checkpoint" under the "File" menu) often, especially before running a cell with a loop for the first time.
3.  If you *do* get stuck in a neverending loop, click on "Kernel... Interrupt".  This will often unstick the loop and allow you to pick up where you left off.  
4.  On a Mac, you might try a "Force Quit" of the Python process, using the Activity Manager.

Now, if you're feeling brave, save your work, change the while loop so that it never ends, and try to recover where you left off.  But be aware that this could cause your computer to freeze or behave erratically, crashing your browser, etc.  Don't panic... it probably won't break your computer permanently.

The neverending loop causes two problems here.  One is with your computer processor, which will be essentially spinning its wheels.  This is called [busy waiting](https://en.wikipedia.org/wiki/Busy_waiting), and your computer will essentially be busy waiting forever.  The other problem is that your loop is printing more and more lines of text into the notebook.  This could easily crash your web browser, which is trying to store and display zillions of lines of numbers.  So be ready for problems!  

### The Euclidean algorithm with a while loop

The **Euclidean Algorithm** is a process of repeated division with remainder.  Beginning with two integers `a` (dividend) and `b` (divisor), one computes quotient `q` and remainder `q` to express `a = qb + r`.  Then `b` becomes the dividend and `r` becomes the divisor, and one repeats.  The repetition continues, and the **last nonzero** remainder is the greatest common divisor of `a` and `b`.

We implement the Euclidean algorithm in a few variations.  The first will be a verbose version, to show the user what happens at every step.  We use a while loop to take care of the repetition.

In [223]:
def Euclidean_algorithm(a,b):
    dividend = a
    divisor = b
    while divisor != 0:   # Recall that != means "is not equal to".
        quotient = dividend // divisor
        remainder = dividend % divisor
        print("{} = {} ({}) + {}".format(dividend, quotient, divisor, remainder))
        dividend = divisor  
        divisor = remainder

In [224]:
Euclidean_algorithm(133, 58)

133 = 2 (58) + 17
58 = 3 (17) + 7
17 = 2 (7) + 3
7 = 2 (3) + 1
3 = 3 (1) + 0


In [225]:
Euclidean_algorithm(1312331323, 58123123)

1312331323 = 22 (58123123) + 33622617
58123123 = 1 (33622617) + 24500506
33622617 = 1 (24500506) + 9122111
24500506 = 2 (9122111) + 6256284
9122111 = 1 (6256284) + 2865827
6256284 = 2 (2865827) + 524630
2865827 = 5 (524630) + 242677
524630 = 2 (242677) + 39276
242677 = 6 (39276) + 7021
39276 = 5 (7021) + 4171
7021 = 1 (4171) + 2850
4171 = 1 (2850) + 1321
2850 = 2 (1321) + 208
1321 = 6 (208) + 73
208 = 2 (73) + 62
73 = 1 (62) + 11
62 = 5 (11) + 7
11 = 1 (7) + 4
7 = 1 (4) + 3
4 = 1 (3) + 1
3 = 3 (1) + 0


This is excellent if we want to know every step of the Euclidean algorithm.  If we just want to know the GCD of two numbers, we can be less verbose.  We carefully return the last nonzero remainder after the while loop is concluded.  This last nonzero remainder becomes the divisor when the remainder becomes zero, and then it would become the dividend in the next (unprinted) line.  That is why we return the (absolute value) of the dividend after the loop is concluded.  You might insert a line at the end of the loop, like `print dividend, divisor, remainder` to help you track the variables.

In [226]:
def GCD(a,b):
    dividend = a # The first dividend is a.
    divisor = b # The first divisor is b.
    while divisor != 0:   # Recall that != means "not equal to".
        quotient = dividend // divisor
        remainder = dividend % divisor
        dividend = divisor  
        divisor = remainder
    return abs(dividend)  #  abs() is used, since we like our GCDs to be positive.

Note that the `return dividend` statement occurs *after* the scope of the while loop.  So as soon as the *divisor* variable equals zero, the funtion `GCD` returns the *dividend* variable and terminates.

In [227]:
GCD(111,27)

3

In [228]:
GCD(111,-27)

3

We can refine our code in a few ways.  First, note that the `quotient` variable is never used!  It was nice in the verbose version of the Euclidean algorithm, but plays no role in finding the GCD.  Our refined code reads
```python
def GCD(a,b):
    dividend = a  
    divisor = b  
    while divisor != 0:   # Recall that != means "not equal to".
        remainder = dividend % divisor
        dividend = divisor  
        divisor = remainder
    return abs(dividend) 
```

Now there are two slick Python tricks we can use to shorten the code.  The first is called **multiple assignment**.  It is possible to set the values of two variables in a single line of code, with a syntax like below.

In [229]:
x,y = 2,3  # Sets x to 2 and y to 3.

This is particular useful for self-referential assignments, because as for ordinary assignment, the right side is evaluated first and then bound to the variables on the left side.  For example, after the line above, try the line below.  Use print statements to see what the values of the variables are afterwards!

In [230]:
x,y = y,x #  Guess what this does!

In [231]:
print("x =",x) # One could use "x = {}".format(x) too.
print("y =",y)

x = 3
y = 2


Now we can use multiple assignment to turn three lines of code into one line of code.  For the `remainder` variable is only used temporarily before its value is given to the `divisor` variable.  Using multiple assignment, the three lines
```python
    remainder = dividend % divisor
    dividend = divisor  
    divisor = remainder
```
can be written in one line,
```python
    dividend, divisor = divisor, dividend % divisor # Evaluations on the right occur before any assignments!
```

Our newly shortened GCD function looks like this.
```python
def GCD(a,b):
    dividend = a  
    divisor = b  
    while divisor != 0:   # Recall that != means "not equal to".
        dividend, divisor = divisor, dividend % divisor
    return abs(dividend)
```

The next trick involves the while loop.  The usual syntax has the form `while <boolean>:`.  But if `while` is followed by a numerical type, e.g. `while <int>:`, then the scope of the while loop will execute as long as the number is nonzero!  Therefore, the line
```python
while divisor != 0:
```
can be replaced by the shorter line
```python
while divisor:
```

This is truly a trick.  It probably won't speed anything up, and it does not make your program easier to read for beginners.  So use it if you prefer communicating with experienced Python programmers!  Here is the whole function again.
```python
def GCD(a,b):
    dividend = a  
    divisor = b  
    while divisor:   # Executes the scope if divisor is nonzero.
        dividend, divisor = divisor, dividend % divisor
    return abs(dividend)
```

The next optimization is a bit more dangerous for beginners, but it works here.  In general, it can be dangerous to operate directly on the arguments to a function.  But in this setting, it is safe, and makes no real difference to the Python interpreter.  Instead of creating new variables called `dividend` and `divisor`, one can manipulate `a` and `b` directly within the function.  If you do this, the GCD function can be shortened to the following.

In [241]:
def GCD(a,b):
    while b:   # I.e., while b != 0.
        a, b = b, a % b
    return abs(a)

In [234]:
# Try it out.  Try it on some big numbers and see how quick it runs!
GCD(2345678198765432456789, 5657829198765389678)

1

This code is essentially optimal, if one wishes to execute the Euclidean algorithm to find the GCD of two integers.  It almost [matches the GCD code in a standard Python library](https://stackoverflow.com/a/18944210).  It might be slightly faster than our original code -- but there is a tradeoff here between execution speed and readability of code.  In this and the following lessons, we often optimize enough for everyday purposes, but not so much that readability is lost.

### Exercises

1.  Modify the `is_prime` function by using a while loop instead of `for j in range(2,n):`.  Hint:  the function should contain the lines `j = 2` and `while j < n:` and `j = j + 1` in various places.  Why might this be an improvement from the for loop?

2.  Modify the `Euclidean_algorithm` function to create a function which returns the *number of steps* that the Euclidean algorithm requires, i.e., the number of divisions-with-remainder.

3.  When $a$ and $b$ are integers, $GCD(a,b) \cdot LCM(a,b) = ab$.  Use this fact to write an LCM-function.  Try to make your function output only integers (not floats) and behave in a good way even if $a,b$ are zero.

4.  How does the `GCD(a,b)` function behave when `a` and/or `b` are zero or negative?  Is this good?

5.  Challenge:  Write a function `approximate_e(n)` which approximates $e$ with a maximum error of $10^{-n}$.  (You can assume $n < 1000$ if necessary.)  Try a while loop and `import mpmath`.  Back up your work often! 

### Solutions
1. `is_prime`: I don't see a reason why this would be prefered in python 3, but in python 2, this would be much better than preloading in the entire *list* `range(2,n)`

```python
def is_prime(n):
    j = 2
    while j < n:
        if n%j == 0:  # is n divisible by j?
            print("{} is a factor of {}.".format(j,n))
            return False
        j += 1
            
    return True
```

2. Step counter:
```python
def euclidean_algorithm_counter(a,b):
    steps = 0
    while b:
        a, b = b, a % b
        steps += 1
    
    # This will include the step where 0 is the last remainder
    print("Euclidean algorithm took", steps, "steps!")
```


3. `LCM(a,b)`:
```python
def LCM(a,b):
    
    # To avoid deviding by zero, just force LCM(0,0) = 0
    if a == b == 0:
        return 0
    
    # We need these because the loop is changing the arguments directly
    original_a = a
    original_b = b
    
    # Get gcd:
    while b:
        a, b = b, a % b
    GCD = abs(a)
    
    # This is the lcm, cast the division to an int first
    # We divide only one of the arguments, in an attempt not to loose information
    return int(original_a/GCD) * original_b
```


4. GCD will give the same answer no matter if one or both of its arguments are negative. If one of the arguments is 0, the result is the nonzero argument. This make sense if we think of 0 being divisible by any nonzero number $x$ in the sense that $\frac{0}{x} = 0$. If both arguments are 0, then the result is zero. This makes sense too.

In [235]:
def is_prime(n):
    j = 2
    while j < n:
        if n%j == 0:  # is n divisible by j?
            print("{} is a factor of {}.".format(j,n))
            return False
        j += 1
            
    return True

In [239]:
def euclidean_algorithm_counter(a,b):
    steps = 0
    while b:
        a, b = b, a % b
        steps += 1
    
    # This will include the step where 0 is the last remainder
    print("Euclidean algorithm took", steps, "steps!")

In [240]:
# Test
euclidean_algorithm_counter(1312331323, 58123123)

Euclidean algorithm took 21 steps!


In [317]:
GCD(-215, 1885)

5

In [300]:
GCD(1,0)

1

In [301]:
GCD(0,0)

0

In [308]:
a, b = 0, 1
a == b == 0

False

In [309]:
def LCM(a,b):
    
    # To avoid deviding by zero, just force LCM(0,0) = 0
    if a == b == 0:
        return 0
    
    # We need these because the loop is changing the arguments directly
    original_a = a
    original_b = b
    
    # Get gcd:
    while b:
        a, b = b, a % b
    GCD = abs(a)
    
    # This is the lcm, cast the division to an int first
    # We divide only one of the arguments, in an attempt not to loose information
    return int(original_a/GCD) * original_b

In [297]:
LCM(13121323, 5813123)

76275864521729

In [298]:
LCM(13123345676573867543567890902876542678912345678901234567802987891323, 581345624657823123)

2543066529982792611700466549785399025111050987614671319233093505798247848516626939904

In [310]:
LCM(0, 0)

0